utils/build_path_prefix_map.mli

```diff
@@ -38,10 +38,37 @@ type map = pair option list
 val encode_map : map -> string
 val decode_map : string -> (map, error_message) result
 
+val find_rewrite : map -> (path -> bool) -> path -> path option
+(** [rewrite_opt map pred path] tries to find a source in [map]
+    that is a prefix of the input [path] that produces a result
+    that satisfies predicate [pred]. If it succeeds,
+    it replaces this prefix with the corresponding target.
+    If it fails, it just returns [None]. *)
+
 val rewrite_opt : map -> path -> path option
 (** [rewrite_opt map path] tries to find a source in [map]
     that is a prefix of the input [path]. If it succeeds,
     it replaces this prefix with the corresponding target.
     If it fails, it just returns [None]. *)
 
 val rewrite : map -> path -> path
+
+val rewrite_exists : map -> path -> path option
+(** [rewrite_exists map path] tries to find a source in [map]
+    that maps to a result that exists in the file system.
+    If so, returns Some result. Otherwise, is there was
+    at least one source that was a prefix of path, it is
+    assumed that the map is deficient and Not_found is raised.
+    If no source prefixes were found, None is returned.
+*)
+
+val matching_dirs : map -> path -> path list
+(** [matching_dirs map absdir] accumulates a list of existing directories,
+  [dirs], that are the result of mapping abstract directory, [absdir],
+  over all the mapping pairs in [map]. The list [dirs] will be in
+  priority order (head as highest priority). If there is success,
+  returns [dirs]. If no source in the map was a prefix of [absdir],
+  returns [[]]. If at least one source in the map was a prefix of
+  [absdir], but none of the mappings were existing directories,
+  raise Not_found.
+*)
```

Comment:
I am not sure that we need to expose those two specialized functions in this module, given that they can easily be defined from the outside using `find_rewrite`. I don't have a strong opinion either and including them here is acceptable to me.

I also wondered if we should include a function that returns the list of possible rewrites, and not the first one only, so that one could implement `matching_dirs` without the false-predicate hack.

Reply:
Note: if we expose a function that returns all possible rewrites in priority order, then we don't actually need `find_rewrite` anymore, the user can define it on top. (There is the slight performance disadvantage of computing all rewrites to use just one, we could use `Seq.t` to make the computation lazy if we wanted to be really fancy, but a list is fine I think.)

This would give the following API:

```ocaml
val rewrite_first : path -> path option
val rewrite_all : path -> path list option   (or: path Seq.t option)

val rewrite : path -> path (* convenience function *)
```

I find the simplicity of the resulting API seductive. Everything else, including showing a warning, can be done in Location or some other module.

Reply:
I adopted your suggestion and I think things turned out quite nicely.